ReportAction text for transactions edited in OldDot

cristipaval opened the pull request:
This is part of the Backward compatibility project.

### Details
In OldDot, it's possible to edit multiple fields at once on transactions. This PR adds support for the `MODIFIEDEXPENSE` report actions that have multiple changes in the `originalMessage`. 


### Fixed Issues
Part of https://github.com/Expensify/Expensify/issues/327756


### Tests
1. The `admin` account has a free workspace and adds the `employee` as a member
2. `employee` creates a money request in workspace chat
3. `admin` signs into OldDot and opens the newly created expense report
4. `admin` edits multiple fields of the transaction (eg. price, category and description)
5. Verify that `admin` sees a new report action in the report history in OldDot, describing the changes
6. Verify that `admin` sees a new report action in the transaction thread in NewDot, describing the changes
7. Verify that `employee` sees a new report action in the transaction thread in NewDot, describing the changes

- [x] Verify that no errors appear in the JS console

### Offline tests
N/A

### QA Steps
Same as Tests

- [x] Verify that no errors appear in the JS console

### PR Author Checklist
<!--
This is a checklist for PR authors. Please make sure to complete all tasks and check them off once you do, or else your PR will not be merged!
-->

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>
<img width="292" alt="Screenshot 2023-10-31 at 14 52 01" src="https://github.com/Expensify/App/assets/18078393/e14c40bf-3f1d-4388-b2f5-f367b2b81bee">
</details>

<details>
<summary>Android: mWeb Chrome</summary>
<img width="275" alt="Screenshot 2023-10-31 at 13 43 24" src="https://github.com/Expensify/App/assets/18078393/514e7513-61b8-41b0-b05d-88353fccaa63">
</details>

<details>
<summary>iOS: Native</summary>

<!-- add screenshots or videos here -->

</details>

<details>
<summary>iOS: mWeb Safari</summary>
<img width="413" alt="Screenshot 2023-10-31 at 13 46 46" src="https://github.com/Expensify/App/assets/18078393/081f11ca-1192-43f7-a6c2-fbf286d9cb90">
</details>

<details>
<summary>MacOS: Chrome / Safari</summary>
<img width="1069" alt="Screenshot 2023-10-30 at 21 04 10" src="https://github.com/Expensify/App/assets/18078393/83b98925-063c-437e-a0da-ae358c948aec">
</details>

<details>
<summary>MacOS: Desktop</summary>
<img width="827" alt="Screenshot 2023-10-30 at 21 06 34" src="https://github.com/Expensify/App/assets/18078393/84e5caa6-08da-450d-a857-e45c7814e29d">
</details>

cristipaval committed:
ReportAction text for transactions edited in OldDot

cristipaval commented:
I'll attach the screenshots soon.

cristipaval committed:
Make Lint happy.

cristipaval committed:
Run prettier.

cristipaval commented:
I added all screenshots except the iOS native, @tgolen. Still working on making my simulator point to my dev env. Discussing it [here](https://expensify.slack.com/archives/C03TQ48KC/p1698753602432659?thread_ts=1693993143.483479&cid=C03TQ48KC)

tgolen commented on src/libs/ReportUtils.js:
```diff
@@ -1864,40 +1866,55 @@ function getModifiedExpenseMessage(reportAction) {
+        if (index === 0) {
+            return acc + value;
+        }
+        return `${acc}. ${value.charAt(0).toUpperCase()}${value.slice(1)}`;
```
I think this looks a bit strange to have a mixture of upper/lower case and also a mix of punctuation:

![image](https://github.com/Expensify/App/assets/1228807/ccd32643-60bd-492c-b904-fa9a7515be8a)

The reason they don't start with a capital is because you should read it like "Cristi changed...". With the addition of these multiple fragments though, it looks off. Here is what I suggest:

1. Each message should start with a lower case
2. Each message should end with a period
3. If there are multiple fragments, then they should be like `changed the date to blah and the amount to blah.` (for two) and `changed the date to blah, the amount to blah, and the description to blah.` (for more than two)

cristipaval replied:
I implemented it as currently is in OldDot.
I really like your suggestion, I'll also ask in the wave4 channel if we want to change it in OldDot as well because it sounds more natural.

cristipaval committed:
Merge remote-tracking branch 'origin/main' into cristi_modified-expense-message

cristipaval committed:
Improve message for MODIFIEDEXPENSE action

cristipaval committed:
Add a safety check

cristipaval committed:
Improve MODIFIEDEXPENSE action message

cristipaval committed:
Improve MODIFIEDEXPENSE action message

cristipaval commented:
@tgolen, I addressed your feedback. The wording is different when the user sets a field that was empty before, changes it, or deletes it.

I know there's a lot of repetitive code that can be improved, but for now I want to see if how the message looks is ok, see below:
<img width="1080" alt="Screenshot 2023-11-08 at 00 33 54" src="https://github.com/Expensify/App/assets/18078393/ecdcc9ef-7766-4168-abb6-9459c4381fd7">

cristipaval committed:
Remove redundancy.

tgolen commented:
I think that looks OK, yeah. Thanks for doing that.

On Tue, Nov 7, 2023 at 3:37 PM Cristi Paval ***@***.***>
wrote:

> @tgolen <https://github.com/tgolen>, I addressed your feedback. The
> wording is different when the user sets a field that was empty before,
> changes it, or deletes it.
>
> I know there's a lot of repetitive code that can be improved, but for now
> I want to see if how the message looks is ok, see below:
> [image: Screenshot 2023-11-08 at 00 33 54]
> <https://user-images.githubusercontent.com/18078393/281200947-ecdcc9ef-7766-4168-abb6-9459c4381fd7.png>
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/Expensify/App/pull/30603#issuecomment-1800308079>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJMAB5PV2QVKK23ZOZKORDYDKZYXAVCNFSM6AAAAAA6WQGPEGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMYTQMBQGMYDQMBXHE>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

cristipaval committed:
Update function doc

cristipaval committed:
Create reusable function

cristipaval committed:
Run prettier

cristipaval committed:
Fix lint.

cristipaval commented:
Ready for a new review, @tgolen.

cristipaval committed:
Update src/libs/ReportUtils.js

Co-authored-by: Tim Golen <tgolen@gmail.com>

cristipaval commented on src/libs/ReportUtils.js:
```diff
@@ -1937,40 +1972,119 @@ function getModifiedExpenseMessage(reportAction) {
+            Localize.translateLocal('common.category'),
+            true,
+        );
+        if (!reportActionOriginalMessage.oldCategory) {
+            setFragments.push(fragment);
+        } else if (!reportActionOriginalMessage.category) {
+            removalFragments.push(fragment);
+        } else {
+            changeFragments.push(fragment);
+        }
```
By moving the logic to a separate module, we'll easily find a way to reuse this. I was reluctant to create 1 more little private function in the huge `ReportUtils.js` file.

cristipaval committed:
Merge remote-tracking branch 'origin/main' into cristi_modified-expense-message

cristipaval committed:
Move logic in a separate unit

cristipaval committed:
Move policy tags logic to PolicyUtils

cristipaval committed:
Better naming

cristipaval committed:
Dry up code.

cristipaval committed:
Run prettier

cristipaval committed:
Fix lint errors

cristipaval committed:
Make it a ts file.

cristipaval committed:
Migrate to TS

cristipaval committed:
Make Lint happy.

cristipaval commented:
Ready again for review, @tgolen

cristipaval commented on src/libs/ReportUtils.js:
```diff
@@ -1880,6 +1880,37 @@ function getProperSchemaForModifiedExpenseMessage(newValue, oldValue, valueName,
+ * @returns {String}
+ */
+
+function getProperLineForModifiedExpenseMessage(prefix, messageFragments) {
```
new module created.

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,229 @@
+    }, prefix);
+}
+
+/**
+ * Get the message for a modified distance request.
+ *
+ * @param newDistance
+ * @param oldDistance
+ * @param newAmount
+ * @param oldAmount
+ */
+
+function getForDistanceRequest(newDistance: string, oldDistance: string, newAmount: string, oldAmount: string) {
```
```suggestion
function getForDistanceRequest(newDistance: string, oldDistance: string, newAmount: string, oldAmount: string) {
```

This is all redundant and doesn't really add any value

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,229 @@
+ * Get the report action message when expense has been modified.
+ *
+ * ModifiedExpense::getNewDotComment in Web-Expensify should match this.
+ * If we change this function be sure to update the backend as well.
+ *
+ * @param reportAction
```
```suggestion
 * If we change this function be sure to update the backend as well.
```

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,229 @@
+ *
+ * @param reportAction
+ */
+function getForReportAction(reportAction: ReportAction) {
```
```suggestion
function getForReportAction(reportAction: ReportAction): string {
```
I think this is how it's done?

tgolen commented on src/libs/PolicyUtils.ts:
```diff
@@ -8,6 +8,22 @@ type MemberEmailsToAccountIDs = Record<string, number>;
+        if (!value) {
+            return;
+        }
+        allPolicyTags = Object.fromEntries(Object.entries(value).filter(([, policyTags]) => !!policyTags));
```
Is `[, policyTags]` right? I've never seen that before so it looks like a mistake. Maybe it's something with TypeScript I am not familiar with.

cristipaval replied:
ugh! fixed.

tgolen commented on src/libs/PolicyUtils.ts:
```diff
@@ -8,6 +8,22 @@ type MemberEmailsToAccountIDs = Record<string, number>;
+    },
+});
+
+function getPolicyTags(policyID: string) {
```
Exporting methods like this (which only return something from an Onyx collection) are a bad pattern. I am trying to get rid of them in https://github.com/Expensify/App/issues/27262. All that needs to happen here is move the Onyx connection to the `ModifiedExpenseMessage` file.

cristipaval replied:
fixed.

cristipaval committed:
Update src/libs/ModifiedExpenseMessage.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

cristipaval committed:
Update src/libs/ModifiedExpenseMessage.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

cristipaval committed:
Update src/libs/ModifiedExpenseMessage.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

cristipaval committed:
Merge remote-tracking branch 'origin/main' into cristi_modified-expense-message

cristipaval committed:
Resolve conflicts after merge.

cristipaval committed:
Move policyTags to ModifiedExpenseMessage

cristipaval committed:
Remove unused function.

cristipaval committed:
Make lint happy.

cristipaval committed:
Better checks.

cristipaval commented:
This is ready again for review, @tgolen

tgolen commented on src/libs/ReportUtils.ts:
```diff
@@ -3650,6 +3496,13 @@ function getReportIDFromLink(url: string | null): string {
+/**
+ * Get the report policyID given a reportID
+ */
+function getReportPolicyID(reportID: string | undefined): string | undefined {
```
Is this the same thing? If so, I think it's more simple and standard.

```suggestion
function getReportPolicyID(reportID?: string)?: string {
```

cristipaval replied:
reverted

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+    if (message === '') {
+        return Localize.translateLocal('iou.changedTheRequest');
+    }
+    message = `${message.substring(1, message.length)}`;
+    return message;
```
```suggestion
    return `${message.substring(1, message.length)}`;
```

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+    }, prefix);
+}
+
+function getForDistanceRequest(newDistance: string, oldDistance: string, newAmount: string, oldAmount: string) {
```
return type declaration

cristipaval replied:
fixed.

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+ * @param valueInQuotes
+ */
+
+function getMessageLine(prefix: string, messageFragments: string[]) {
```
return type declaration

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+ * @param valueInQuotes
+ * @param shouldConvertToLowercase
+ */
+
```
Remove extra line

cristipaval replied:
fixed.

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+
+/**
+ * Builds the partial message fragment for a modified field on the expense.
+ *
+ * @param newValue
+ * @param oldValue
+ * @param valueName
+ * @param valueInQuotes
+ * @param shouldConvertToLowercase
```
Param docs aren't needed for TS

cristipaval replied:
removed

tgolen commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+ * @param valueName
+ * @param valueInQuotes
+ */
+
```
Extra line

cristipaval replied:
removed

cristipaval committed:
Update src/libs/ReportUtils.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

cristipaval committed:
Update src/libs/ModifiedExpenseMessage.ts

Co-authored-by: Tim Golen <tgolen@gmail.com>

cristipaval committed:
Improve function doc.

cristipaval committed:
Add return type declaration

cristipaval commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,238 @@
+ * @param newValue
+ * @param oldValue
+ * @param valueName
+ * @param valueInQuotes
```
removed

cristipaval commented:
ready again, @tgolen

tgolen commented on src/libs/ReportUtils.ts:
```diff
@@ -3650,6 +3496,13 @@ function getReportIDFromLink(url: string | null): string {
+/**
+ * Get the report policyID given a reportID
+ */
+function getReportPolicyID(reportID?: string): string | undefined {
```
I think the function's return type can also be changed to `?: string`?

cristipaval replied:
I tried to change the return type as you suggested but it doesn't work, I get syntactic errors, see [here](https://github.com/Expensify/App/pull/30603#discussion_r1409590041)

tgolen reviewed (commented):
Great, thanks for the updates. I noticed this one last thing. Also, there are some conflicts that creeped in.

cristipaval committed:
Merge remote-tracking branch 'origin/main' into cristi_modified-expense-message

cristipaval commented:
@situchan, the screenshots from the description body of this PR might be deprecated as we changed the approach during a few rounds of reviews done by @tgolen
[Here](https://github.com/Expensify/App/pull/30603#issuecomment-1800308079) you can see how the message should be structured.

cristipaval commented:
@situchan, let me know if you need help generating the `MODIFIEDEXPENSE` reportAction with multiple edited fields.

situchan commented:
> @situchan, let me know if you need help generating the `MODIFIEDEXPENSE` reportAction with multiple edited fields.

yes please

cristipaval commented:
> > @situchan, let me know if you need help generating the `MODIFIEDEXPENSE` reportAction with multiple edited fields.
> 
> yes please

Did you try steps 3 and 4 from the Tests? Did they work?

situchan commented:
Thanks. I now got the issue. Reviewing

situchan commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,224 @@
+        getMessageLine(`\n${Localize.translateLocal('iou.changed')}`, changeFragments) +
+        getMessageLine(`\n${Localize.translateLocal('iou.set')}`, setFragments) +
+        getMessageLine(`\n${Localize.translateLocal('iou.removed')}`, removalFragments);
+    if (message === '') {
+        return Localize.translateLocal('iou.changedTheRequest');
+    }
+    return `${message.substring(1, message.length)}`;
```
```suggestion

    return message ? message.substring(1) : Localize.translateLocal('iou.changedTheRequest');
```

situchan reviewed (commented):
Code looks good.

Do you think it's worth adding automated tests for various cases?

situchan commented:
Should we support attendees change as well?

OD:

<img width="894" alt="Screenshot 2023-12-01 at 2 30 14 PM" src="https://github.com/Expensify/App/assets/108292595/d9a49feb-ed3e-4045-8356-aaadc832c732">

ND:

<img width="728" alt="Screenshot 2023-12-01 at 2 30 27 PM" src="https://github.com/Expensify/App/assets/108292595/98d2a5af-9ca1-4680-af02-0f563394eecf">

Action:

<img width="912" alt="Screenshot 2023-12-01 at 2 30 53 PM" src="https://github.com/Expensify/App/assets/108292595/e4498b7f-fc2a-450c-9781-7a3a5e0fe986">

cristipaval commented:
we'll definitely have to support attendee changes. I'm just wondering if this should be part of [Attendee tracking project](https://github.com/Expensify/Expensify/issues/306605)

cristipaval commented:
> we'll definitely have to support attendee changes. I'm just wondering if this should be part of [Attendee tracking project](https://github.com/Expensify/Expensify/issues/306605)

I'd say let's ignore the attendees for now and leave it for when we introduce attendee tracking in NewDot. What this PR has is better than what we currently have in the App. If we think that this is something that really needs to be done before, I can create a follow-up PR. I just wouldn't keep changing the current one.

Having said that, I'm going to write unit tests and then push to get this one merged.

cristipaval committed:
Merge remote-tracking branch 'origin/main' into cristi_modified-expense-message

cristipaval committed:
Fix conflicts after merge

cristipaval committed:
RUn prettier

cristipaval committed:
Performance test for ModifiedExpenseMessage

cristipaval committed:
Fix typecheck failures

cristipaval committed:
Unit test for ModifiedExpenseMessage

cristipaval committed:
More tests.

cristipaval committed:
One more test

cristipaval committed:
One more test

cristipaval committed:
3 more tests

cristipaval committed:
3 more tests

cristipaval committed:
Make lint happy an run prettier

cristipaval commented:
This is again ready for review @situchan!

dylanexpensify commented:
bump @situchan for review!

cristipaval committed:
Merge remote-tracking branch 'origin/main' into cristi_modified-expense-message

cristipaval committed:
Add changes from main

situchan commented:
I will retest and checklist today

cristipaval commented:
> I will retest and checklist today

Please do; I want to merge this one, as conflicts are coming in.

situchan commented:
oldDot:
`You modified expense 'Merchant'. New merchant is 'Merchant2' (previously 'Merchant'). New date is '2023-12-01' (previously '2023-12-07'). New amount is '$4.00' (previously '$6.00'). New reimbursability status is 'No'. New description is 'test2' (previously 'test'). Added situchandrashil7+23@gmail.com as an attendee.`

this branch:
`changed the amount to $4.00 (previously $6.00), the description to "test2" (previously "test"), the date to 2023-12-01 (previously 2023-12-06), and the merchant to "Merchant2" (previously "Merchant").`

staging:
`changed the amount to $4.00 (previously $6.00)`


I noticed that date doesn't match (In above example, `2023-12-07` in oldDot, `2023-12-06` in newDot).
But the root cause already exists in production: #33248
So I consider this out of scope

situchan commented:
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I verified tests pass on **all platforms** & I tested again on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

<img width="365" alt="android" src="https://github.com/Expensify/App/assets/108292595/f0eb6a68-21fc-4dfe-9a14-bb32c1b555cb">


</details>

<details>
<summary>Android: mWeb Chrome</summary>

<img width="365" alt="mchrome" src="https://github.com/Expensify/App/assets/108292595/d131c98a-e0ae-449b-9afa-2ca7ca5f6137">


</details>

<details>
<summary>iOS: Native</summary>

![ios](https://github.com/Expensify/App/assets/108292595/941f609b-00b0-401f-8840-4a927af4b225)


</details>

<details>
<summary>iOS: mWeb Safari</summary>

![msafari](https://github.com/Expensify/App/assets/108292595/17bb35a1-42e6-419b-8544-ad83adc71473)


</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

<img width="1152" alt="web" src="https://github.com/Expensify/App/assets/108292595/154856a0-ec00-49e1-b3be-6bc4a59570d3">


</details>

<details>
<summary>MacOS: Desktop</summary>

<img width="1185" alt="desktop" src="https://github.com/Expensify/App/assets/108292595/d046e565-a494-469f-ba1b-ff1584632e6d">


</details>

situchan reviewed (approved):
:shipit:

stitesExpensify reviewed (approved):
LGTM!

OSBotify commented:
:hand: This PR was not deployed to staging yet because QA is ongoing. It will be automatically deployed to staging after the next production release.

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7284216530) to staging by https://github.com/stitesExpensify in version: 1.4.15-0 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|success ✅
🕸 web 🕸|success ✅

OSBotify commented:
🚀 [Deployed](https://github.com/Expensify/App/actions/runs/7296457636) to production by https://github.com/jasperhuangg in version: 1.4.15-5 🚀

platform | result
---|---
🤖 android 🤖|success ✅
🖥 desktop 🖥|success ✅
🍎 iOS 🍎|failure ❌
🕸 web 🕸|success ✅

jjcoffee commented on src/libs/ModifiedExpenseMessage.ts:
```diff
@@ -0,0 +1,226 @@
+        return '';
+    }
+    const reportActionOriginalMessage = reportAction.originalMessage as ExpenseOriginalMessage | undefined;
+    const policyID = ReportUtils.getReportPolicyID(reportAction.reportID) ?? '';
```
`MODIFIEDEXPENSE` report actions don't have the `reportID` property, so we fail to get the `policyTagListName` here. This caused https://github.com/Expensify/App/issues/33493.
